documentation/specs/src/economics/fee-system.md

```diff
@@ -1,7 +1,143 @@
+- Moving insufficient funds would incentivize the block proposer to include transactions for which fees cannot be paid. Since the block proposer knows the balances of the involved addresses at block creation time and given the strategy of placing wrapper txs first, this would constitute malicious behavior by the proposer
+
+By discarding the transaction without paying fees instead we avoid these pitfalls. This logic implies that the strategy of placing wrapper transactions before any decrypted tx in the block will be reinforced. It might look like a contradiction to what was said before, a wrapper transaction included in a block will not pay fees for the resources it acquired, but this is not true:
+
+- An address might have no funds at all making it impossible to perform any payment
+- The only actor with an interest in managing block resources is the block proposer who's aware of the balances of the involved addresses
+
+To support the in-protocol fee payment mechanism we need to update the `Header` struct to carry the `ProposerAddress`:
+
+```rust
+/// The data from Tendermint header
+/// relevant for Anoma storage
+#[derive(Clone, Debug, BorshSerialize, BorshDeserialize)]
+pub struct Header {
+    /// Merkle root hash of block
+    pub hash: Hash,
+    /// Timestamp associated to block
+    pub time: DateTimeUtc,
+    /// Hash of the addresses of the next validator set
+    pub next_validators_hash: Hash,
+    /// Address of the block proposer
+    pub proposer_address: Address
+}
+```
+
+From this address, it is then possible to derive the relative Namada address for the payment.
+
+The `Fee` field of `WrapperTx` is defined as follows:
+
+```rust
+pub struct Fee {
+    /// amount of the fee
+    pub amount: Amount,
+    /// address of the token
+    pub token: Address,
+}
+```
+
+The signer of the wrapper transaction defines the token in which fees must be paid among those available in the token whitelist. At the same time, he also sets the amount which must meet the minimum price per gas unit for that token $GP_{min}$ (also defined in the whitelist). The difference between the minimum and the actual value set by the submitter represents the incentive for the block proposer to prefer the inclusion of this transaction over other ones.
+
+The block proposer can check the validity of these two parameters while constructing the block. These validity checks are also replicated in `process_proposal` and `mempool_check`. In case a transaction with invalid parameters ended up in a block, it would be discarded without paying any fee (as already explained earlier in this document). The block proposer and the `process_proposal` function should also cache the available funds of every fee-paying address involved in the block: this is because a signer might submit more than one transaction per block and the check on funds should take into consideration the updated value of the unshielded balance.
```

Comment:
correct proposers should never create blocks with invalid transactions, and such block proposals should be rejected, if possible